app/frontend/src/features/messages/groupchats/GroupChatView.tsx

```diff
@@ -29,6 +29,15 @@ export default function GroupChatView({
     await fetchMessages();
   };
 
+  const leaveGroupChat = async () => {
+    try {
+      await service.conversations.leaveGroupChat(groupChat.groupChatId);
+      handleClose();
+    } catch (error) {
+      setError(error.message);
```

Comment:
I just realised it is possible that either of these `setX` calls (`handleClose` calls `setGroupChats` in parent tab component) could happen after the component has been unmounted from the user switching tabs while things are still loading, which is not ideal as that leads to app memory leaks as indicated from the following famous React warning I'm getting:
```
index.js:1 Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
```

Reply:
The best way I can think of is having to wrap these in a check that the component is still mounted before you call `setX` with a ref and `useLayoutEffect` with an empty dep that flicks the flag to true/false depending on the component's mount state.

Reply:
We are probably not the first to encounter this. Is there some best practice described for this?

Reply:
Indeed, this is a [good blog post](https://www.robinwieruch.de/react-warning-cant-call-setstate-on-an-unmounted-component) I found about it referred to by Dan in the React core team. It seems like I was on the right track - the post references class components, but the idea can be implemented with what I mentioned in a function component with hooks.

Reply:
This would work, right?

```
const [mounted, setMounted] = useState(true);
useEffect(() => () => setMounted(false), []);
```

Reply:
The main problem I can see is that `useState` calls are batched so there's still a chance that a Promise resolves/rejects and tries to set state even though the component has already unmounted, whereas `mounted.current = false` from a ref would be immediate as it's setting the variable directly.

Also I believe when `useState` is initially called, the component is still initialising so it's technically not mounted until `useEffect` is called, so again a small chance that state could be set if the Promise resolves/rejects before the component is actually mounted.

Reply:
So I was thinking more like this:
```typescript
function useIsMounted() {
  const isMounted = useRef(false);

  useEffect(() => {
    isMounted.current = true;
    return () => (isMounted.current = false);
  }, []);

 return isMounted.current;
}
```

Might even worth turn this into a `useIsMounted` custom hook since this will probably come up each time we need to setState in any async operations?

Reply:
I moved a lot of state to the model.

Reply:
I think your example of keeping a mounted state was for me the proverbial drop that went over the bucket. The use of redux and thunks already made a big mess of the components, so I moved most of it out now.